src/equations/compressible_euler_1d.jl

```diff
@@ -150,19 +154,20 @@ function initial_condition_weak_blast_wave(x, t,
                                            equations::CompressibleEulerEquations1D)
     # From Hennemann & Gassner JCP paper 2020 (Sec. 6.3)
     # Set up polar coordinates
-    inicenter = SVector(0.0)
+    RealT = eltype(x)
+    inicenter = SVector(0)
     x_norm = x[1] - inicenter[1]
     r = abs(x_norm)
     # The following code is equivalent to
     # phi = atan(0.0, x_norm)
     # cos_phi = cos(phi)
     # in 1D but faster
-    cos_phi = x_norm > 0 ? one(x_norm) : -one(x_norm)
+    cos_phi = x_norm > 0 ? 1 : -1
 
     # Calculate primitive variables
-    rho = r > 0.5 ? 1.0 : 1.1691
-    v1 = r > 0.5 ? 0.0 : 0.1882 * cos_phi
-    p = r > 0.5 ? 1.0 : 1.245
+    rho = r > 0.5f0 ? 1 : convert(RealT, 1.1691)
+    v1 = r > 0.5f0 ? 0 : convert(RealT, 0.1882) * cos_phi
+    p = r > 0.5f0 ? 1 : convert(RealT, 1.245)
```

Comment:
I don't understand this. Let's assume `RealT == Float64`. Then, you get a type instability like

```julia
julia> foo(r) = r > 0.5f0 ? 1 : 1.1691
foo (generic function with 1 method)

julia> @code_warntype foo(1.0)
MethodInstance for foo(::Float64)
  from foo(r) @ Main REPL[1]:1
Arguments
  #self#::Core.Const(foo)
  r::Float64
Body::Union{Float64, Int64}
1 ─ %1 = (r > 0.5f0)::Bool
└──      goto #3 if not %1
2 ─      return 1
3 ─      return 1.1691
```

Reply:
Yes and it is returning `prim2cons(SVector(rho, v1, p), equations)` at the end. Calling `prim2cons` will promote the final result to the type we want even though we get integers here. Please check here
https://github.com/trixi-framework/Trixi.jl/pull/1909/files#r1594411469

Reply:
I don't know all the technical details behind this discussion, but I intuitively I'd favor being type stable also within in the function, especially when it's as easy as changing a `1` to a `1.0f0`.

I believe that otherwise the function will become unnecessarily slow, even though the outcome is the same, since it will have to add branching etc. to handle the type-unstable part of the function

Reply:
Yes, please - let's also keep type stability inside of functions like this one

Reply:
It is not rigorous. We also have to consider the speed of integer operations v.s. floating-point operations on both CPUs and GPUs. Generally, the former is faster than the later. If we obtain much more integer results (`rho, v1, p`) in this function, then `prim2cons` function will perform integer operations most of the time, thus we can achieve a speedup. So whether to use integer operations here depends on the likelihood of branching into 2 and 3 as below. Who have any idea about this?
```
Body::Union{Float64, Int64}
1 ─ %1 = (r > 0.5f0)::Bool
└──      goto #3 if not %1
2 ─      return 1
3 ─      return 1.1691
```

Reply:
The type instability overhead is dominating and exceeding any possible speedup from integer operations (that we don't get here in the later steps):

```julia
julia> using BenchmarkTools, Trixi

julia> function foo(x, equations)
           RealT = eltype(x)
           inicenter = SVector(0)
           x_norm = x[1] - inicenter[1]
           r = abs(x_norm)
           cos_phi = x_norm > 0 ? 1 : -1
           rho = r > 0.5f0 ? 1 : convert(RealT, 1.1691)
           v1 = r > 0.5f0 ? 0 : convert(RealT, 0.1882) * cos_phi
           p = r > 0.5f0 ? 1 : convert(RealT, 1.245)
           
           return prim2cons(SVector(rho, v1, p), equations)
       end
foo (generic function with 1 method)

julia> function bar(x, equations)
           RealT = eltype(x)
           inicenter = SVector(0)
           x_norm = x[1] - inicenter[1]
           r = abs(x_norm)
           cos_phi = x_norm > 0 ? 1 : -1
           rho = r > 0.5f0 ? one(RealT) : convert(RealT, 1.1691)
           v1 = r > 0.5f0 ? zero(RealT) : convert(RealT, 0.1882) * cos_phi
           p = r > 0.5f0 ? one(RealT) : convert(RealT, 1.245)
           
           return prim2cons(SVector(rho, v1, p), equations)
       end
bar (generic function with 1 method)

julia> equations = CompressibleEulerEquations1D(1.4)
┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
│ CompressibleEulerEquations1D                                                                     │
│ ════════════════════════════                                                                     │
│ #variables: ………………………………………………… 3                                                                │
│ │ variable 1: …………………………………………… rho                                                              │
│ │ variable 2: …………………………………………… rho_v1                                                           │
│ │ variable 3: …………………………………………… rho_e                                                            │
└──────────────────────────────────────────────────────────────────────────────────────────────────┘

julia> @benchmark foo($(SVector(1.0)), $equations)
BenchmarkTools.Trial: 10000 samples with 192 evaluations.
 Range (min … max):  511.281 ns …   7.372 μs  ┊ GC (min … max): 0.00% … 91.35%
 Time  (median):     515.844 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   530.094 ns ± 118.157 ns  ┊ GC (mean ± σ):  0.49% ±  2.11%

  ▄█▇▅▃▃▂▁  ▂▃▂▂▂▃▂▂▂▂     ▁▁▁▁▁                                ▂
  ████████▇▇█████████████████████████▇███▇▇▇▇▆▇▇▅▆▅▅▅▅▅▅▅▅▄▅▅▅▄ █
  511 ns        Histogram: log(frequency) by time        624 ns <

 Memory estimate: 160 bytes, allocs estimate: 5.

julia> @benchmark bar($(SVector(1.0)), $equations)
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  1.958 ns … 17.167 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     2.000 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   2.078 ns ±  0.630 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █▇▁                                                        ▁
  ███▇▁▁▁▁▁▁▃▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▃▁▁▁▁▁▁▁▁▁▁▁▁▁▆▄▃▁▃▅ █
  1.96 ns      Histogram: log(frequency) by time     5.79 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

Reply:
Good.